# methylpy
Welcome to the home page of methylpy, a pyhton-based analysis pipeline for
* (single-cell) (whole-genome) bisulfite sequencing data
* (single-cell)  NOMe-seq data

# Note
* [tutorial](https://github.com/yupenghe/methylpy/blob/methylpy/tutorial.md) is being written
* Current version methylpy has major difference compared to previous version.
Please checkout this document and [tutorial](https://github.com/yupenghe/methylpy/blob/methylpy/tutorial.md)
for details.

# What can methylpy do?
#### Processing bisulfite sequencing data and NOMe-seq data
* fast and flexible pipeline for both single-end and paired-end data
* all the way from raw reads (fastq) to methylation state and/or open chromatin readouts
* also support getting readouts from alignment (BAM file)
* including options for read trimming, quality filter and PCR duplicate removal
* accept compressed input and generate compressed output
* support post-bisulfite adaptor tagging (PBAT) data

#### Calling differentially methylated regions (DMRs)
* DMR calling at single cytosine level
* support comparison across 2 or more samples/groups
* conservative and accurate
* useful feature for dealing with low-coverage data by combining data of adjacent cytosines

# What you want to do
* [Install methylpy](#install-methylpy)
* [Use methylpy for data processing](#use-methylpy-for-data-processing)
* [Use methylpy for calling DMRs](#use-methylpy-for-calling-dmrs)
* [Optional functions for data processing](#optional-functions-for-data-processing)
* [Cite methylpy](#cite-methylpy)

run `methylpy -h` to get a list of functions.

# Install methylpy
#### Step 1 - Download methylpy and set up environment variable
Enter the directory where you would like to install methylpy and run
```
git clone https://github.com/yupenghe/methylpy.git
```
Next, the methylpy folder needs to be included in the python search path. Also, methylpy executable will easy to use if its path is included in `PATH`. To do these, if the path of methylpy (not the methylpy/methylpy folder) is /YOUR/PATH/methylpy/, please include the below code in the $HOME/.bashrc file:
```
export PYTHONPATH=/YOUR/PATH/methylpy/:$PYTHONPATH
export PATH=/YOUR/PATH/methylpy/bin/:$PATH
```
and then do
```
source $HOME/.bashrc
```
If the below command gives no error, the setup is successful.
```
methylpy
```

#### Step 2 - Install dependencies
methylpy is written in python so obviously python2/3 needs to be installed.
methylpy also depends on two python modules, [numpy](http://www.numpy.org/) 
and [scipy](https://www.scipy.org/).
The easiest way to resolve these dependencies is to install [anaconda](https://www.anaconda.com/download/)

In addition, some features of methylpy depend on several publicly available tools
* [cutadapt](http://cutadapt.readthedocs.io/en/stable/installation.html) (>=1.12) for raw read trimming
* [bowtie](http://bowtie-bio.sourceforge.net/index.shtml) and/or 
[bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) for alignment
* [samtools](https://github.com/samtools/samtools) (>=1.3) for alignment results manipulation
* [Picard](https://broadinstitute.github.io/picard/index.html) (>=2.10.8) for removal of PCR duplicates

Lastly, if paths to cutadapt, bowtie/bowtie2 and samtools are included in `PATH` variable,
methylpy can run these tools directly. Otherwise, the paths have to be passed to methylpy as augments. 
Path to Picard needs to be passed to methylpy as a parameter to run PCR duplicate removal.

#### Optional step - Compile rms.cpp
DMR finding requires an executable `methylpy/methylpy/run_rms_tests.out`, which was compiled from
C++ code `methylpy/methylpy/rms.cpp`. In most cases, the precompiled file can be used directly. To
test this, simply run execute `methylpy/methylpy/run_rms_tests.out`. If help page shows, recompiling
is not required. If error turns up, the executable needs to be regenerated by compiling `rms.cpp` and
this step requires [GSL](https://www.gnu.org/software/gsl/) installed correctly. In most linux operating 
system, the below commands will do the job
```
cd methylpy/methylpy/
g++ -O3 -l gsl -l gslcblas -o run_rms_tests.out rms.cpp
```
In Ubuntu (>=16.04), please try the below commands first.
```
cd methylpy/methylpy/
g++ -o run_rms_tests.out rms.cpp `gsl-config --cflags â€”libs`
```

# Use methylpy for data processing
Please see [methylpy tutorial](https://github.com/yupenghe/methylpy/blob/methylpy/tutorial.md)
for more details.

#### Step 1 - Build converted genome reference
Build bowtie/bowtie2 index for converted genome. Run `methylpy build-reference -h`
to get more information. An example of building mm10 mouse reference index:

```
/gale/netapp/home/yupeng/methylpy_github/bin/methylpy \
    build-reference \
	--input-files mm10_bt2/mm10.fa \
	--output-prefix mm10_bt2/mm10 \
	--bowtie2 True
```

#### Step 2 - Process bisulfite sequencing and NOMe-seq data 
Function `single-end-pipeline` is For processing single-end data. Run 
`methylpy single-end-pipeline -h` to get help information. Below code
is an example of using methylpy to process single-end bisulfite sequencing
data. For processing NOMe-seq data, please use `num_upstr_bases=1` to include
one base upstream cytosine as part of cytosine sequence context, which can be
used to tease out GC sites.

```
/gale/netapp/home/yupeng/methylpy_github/bin/methylpy \
    single-end-pipeline \
	--read-files raw/mESC_R1.fastq.gz \
	--sample mESC \
    --forward-ref mm10_bt2/mm10_f \
    --reverse-ref mm10_bt2/mm10_r \
    --ref-fasta mm10_bt2/mm10.fa \
    --num-procs 8 \
    --remove-clonal True \
    --path-to-picard="picard/"
```

An command example for processing paired-end data.
Run `methylpy paired-end-pipeline -h` to get more information. 

```
/gale/netapp/home/yupeng/methylpy_github/bin/methylpy \
    paired-end-pipeline \
    --read1-files raw/mESC_R1.fastq.gz \
    --read2-files raw/mESC_R2.fastq.gz \
    --sample mESC \
    --forward-ref mm10_bt2/mm10_f \
    --reverse-ref mm10_bt2/mm10_r \
    --ref-fasta mm10_bt2/mm10.fa \
    --num-procs 8 \
    --remove-clonal True \
    --path-to-picard="picard/"
```

#### Output format
Output file(s) are (compressed) tab-separated text file(s) in allc format. "allc" stands
for all cytosine (C). Each row in an allc file corresponds to one cytosine in the genome. 
An allc file contain 7 columns and no header:

|index|column name|example|note|
|:----:|:-----:|:-----:|:---:|
|1|chromosome|12|with no "chr"|
|2|position|18283342|1-based|
|3|strand|+|either + or -|
|4|sequence context|CGT|can be more than 3 bases|
|5|mc|18|count of reads supporting methylation|
|6|cov|21|read coverage|
|7|methylyate|1|indicator of significant methylation|

# Using methylpy for calling DMRs
This function will take a list of compressed/uncompressed allc files (output files from methylpy pipeline) as input
and look for DMRs. Help information of this function is available via running `methylpy DMRfind -h`. 

Below is the code of an example of calling DMRs for CG methylation between two samples, 
`AD_HT` and `AD_IT` on chromosome 1 through 5 using 8 processors.

```
/gale/netapp/home/yupeng/methylpy_github/bin/methylpy \
	DMRfind \
	--allc-files allc/allc_AD_HT.tsv.gz allc/allc_AD_IT.tsv.gz \
	--samples AD_HT AD_IT \
	--mc-type "CGN" \
	--chroms 1 2 3 4 5 \
	--num-procs 8 \
	--output-prefix DMR_HT_IT
```
Please see [methylpy tutorial](https://github.com/yupenghe/methylpy/blob/methylpy/tutorial.md) for details.

# Optional functions for data processing
#### Extract cytosine methylation state from BAM file
This function allows users to get cytosine methylation state (allc file) from alignment file (BAM file).
It is part of the data processing pipeline which is especially useful for getting the allc file from
alignment file from other methylation data pipelines like bismark. Run `methylpy call-methylation-state -h`
to get help information. Below is an example of running this function.

```
/gale/netapp/home/yupeng/methylpy_github/bin/methylpy \
    call-methylation-state \
	--input-file mESC_processed_reads_no_clonal.bam \
	--paired-end True \
	--sample mESC \
	--ref-fasta mm10_bt2/mm10.fa \
	--num-procs 8
```

#### Quality filter for bisulfite sequencing reads
Sometimes, we want to filter out reads that cannot be mapped confidently or are likely from 
under-converted DNA fragments. This can be done using the `bam-quality-filter` function.
See `methylpy bam-quality-filter -h` for parameter inforamtion. 

For example, below command can be used to filter out reads with less than 30 MAPQ score
(poor alignment) and with mCH level greater than 0.7 (under-conversion) if the reads contain
enough (at least 3) CH sites.

```
/gale/netapp/home/yupeng/methylpy_github/bin/methylpy \
    bam-quality-filter \
	--input-file mESC_processed_reads_no_clonal.bam \
	--output-file mESC_processed_reads_no_clonal.filtered.bam \
	--ref-fasta mm10_bt2/mm10.fa \
	--quality-cutoff 30 \
	--min-num-ch 3 \
	--max-mch-level 0.7 \
	--buffer-line-number 100
```

# Cite methylpy
If you use methylpy, please cite
```
Matthew D. Schultz, Yupeng He, John W.Whitaker, Manoj Hariharan, Eran A. Mukamel,
Danny Leung, Nisha Rajagopal, Joseph R. Nery, Mark A. Urich, Huaming Chen, Shin Lin, 
Yiing Lin, Bing Ren, Terrence J. Sejnowski, Wei Wang, Joseph R. Ecker. 
Human Body Epigenome Maps Reveal Noncanonical DNA Methylation Variation.
Nature. 523(7559):212-216, 2015 Jul.
```
