# methylpy
Welcome to the home page of methylpy, a pyhton-based analysis pipeline for
* (single-cell) (whole-genome) bisulfite sequencing data
* (single-cell)  NOMe-seq data

# Note
* Currently, methylpy only supports python2.7 but support for python3 is coming.
* methylpy has major changes compared to previous version. Please double check whether your code is compatible.

# What can methylpy do?
#### Processing bisulfite sequencing data and NOMe-seq data
* fast and flexible pipeline for both single-end and paired-end data
* all the way from raw reads (fastq) to methylation state and/or open chromatin readouts
* also support getting readouts from alignment (BAM file)
* including options for read trimming, quality filter and PCR duplicate removal
* accept compressed input and generate compressed output
* support post-bisulfite adaptor tagging (PBAT) data

#### Calling differentially methylated regions (DMRs)
* DMR calling at single cytosine level
* support comparison across 2 or more samples/groups
* conservative and accurate
* useful feature for dealing with low-coverage data by combining data of adjacent cytosines

# What you want to do
* [Installing methylpy](#installing-methylpy)
* [Using methylpy for data processing](#using-methylpy-for-data-processing)
* [Using methylpy for calling DMRs](#using-methylpy-for-calling-dmrs)
* [Optional steps for data processing](#optional-steps-for-data-processing)
run `methylpy -h` to get a list of functions.

# Installing methylpy
#### step 1 - download methylpy and set up environment variable
Enter the directory where you would like to install methylpy and run
```
git clone https://github.com/yupenghe/methylpy.git
```
Next, the methylpy folder needs to be included in the python search path. Also, methylpy executable will easy to use if its path is included in `PATH`. To do these, if the path of methylpy (not the methylpy/methylpy folder) is /YOUR/PATH/methylpy/, please include the below code in the $HOME/.bashrc file:
```
export PYTHONPATH=/YOUR/PATH/methylpy/:$PYTHONPATH
export PATH=/YOUR/PATH/methylpy/bin/:$PATH
```
and then do
```
source $HOME/.bashrc
```
If the below command gives no error, the setup is successful.
```
methylpy
```

#### step 2 - install dependencies
methylpy is written in python2 so obviously python2 needs to be installed.
methylpy also depends on two python modules, [numpy](http://www.numpy.org/) 
and [scipy](https://www.scipy.org/).
The easiest way to resolve these dependencies is to install [anaconda](https://www.anaconda.com/download/)

In addition, some features of methylpy depend on several publicly available tools
* [cutadapt](http://cutadapt.readthedocs.io/en/stable/installation.html) (>=1.12) for raw read trimming
* [bowtie](http://bowtie-bio.sourceforge.net/index.shtml) and/or [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) for alignment
* [samtools](https://github.com/samtools/samtools) (>=1.3) for alignment results manipulation
* [Picard](https://broadinstitute.github.io/picard/index.html) (>=2.10.8) for removal of PCR duplicates

Lastly, if paths to cutadapt, bowtie/bowtie2 and samtools are included in `PATH` variable,
methylpy can run these tools directly. Otherwise, the paths have to be passed to methylpy as augments. 
Path to Picard needs to be passed to methylpy as a parameter to run PCR duplicate removal.

#### optional step - compile rms.cpp
DMR finding requires an executable `methylpy/methylpy/run_rms_tests.out`, which was compiled from
C++ code `methylpy/methylpy/rms.cpp`. In most cases, the precompiled file can be used directly. To
test this, simply run execute `methylpy/methylpy/run_rms_tests.out`. If help page shows, recompiling
is not required. If error turns up, the executable needs to be regenerated by compiling `rms.cpp` and
this step requires [GSL](https://www.gnu.org/software/gsl/) installed correctly. In most linux operating 
system, the below commands will do the job
```
cd methylpy/methylpy/
g++ -O3 -l gsl -l gslcblas -o run_rms_tests.out rms.cpp
```
In Ubuntu (>=16.04), please try the below commands first.
```
cd methylpy/methylpy/
g++ -o run_rms_tests.out rms.cpp `gsl-config --cflags â€”libs`
```

# Using methylpy for data processing
Please see [methylpy tutorial](https://github.com/yupenghe/methylpy/blob/methylpy/tutorial.md) for details.

#### step 1 - Building converted genome reference

`methylpy build-reference -h`

#### step 2 - processing bisulfite sequencing and NOMe-seq data 
For single-end data, `methylpy single-end-pipeline -h`
For paired-end data, `methylpy paired-end-pipeline -h`

Output file(s) are (compressed) tab-separated text file(s) in allc format. "allc" stands
for all cytosine (C). Each row in an allc file corresponds to one cytosine in the genome. 
An allc file contain 7 columns and no header:

|index|column name|example|note|
|:----:|:-----:|:-----:|:---:|
|1|chromosome|12|with no "chr"|
|2|position|18283342|1-based|
|3|strand|+|either + or -|
|4|sequence context|CGT|can be more than 3 bases|
|5|mc|18|count of reads supporting methylation|
|6|cov|21|read coverage|
|7|methylyate|1|indicator of significant methylation|

# Using methylpy for calling DMRs
Please see [methylpy tutorial](https://github.com/yupenghe/methylpy/blob/methylpy/tutorial.md) for details.
`methylpy DMRfind -h`

# Optional steps for data processing
#### Quality filter for bisulfite sequencing reads
Sometimes, 
`methylpy bam-quality-filter -h`

#### Extracting methylation state from BAM file
This module allows users to get 
`methylpy call-methylation-state -h`
